<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Input</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>
<script>// Hide empty <a> tag within highlighted CodeBlock for screen reader accessibility (see https://github.com/jgm/pandoc/issues/6352#issuecomment-626106786) -->
// v0.0.1
// Written by JooYoung Seo (jooyoung@psu.edu) and Atsushi Yasumoto on June 1st, 2020.

document.addEventListener('DOMContentLoaded', function() {
  const codeList = document.getElementsByClassName("sourceCode");
  for (var i = 0; i < codeList.length; i++) {
    var linkList = codeList[i].getElementsByTagName('a');
    for (var j = 0; j < linkList.length; j++) {
      if (linkList[j].innerHTML === "") {
        linkList[j].setAttribute('aria-hidden', 'true');
      }
    }
  }
});
</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Input</h1>



<div id="input-from-scratch" class="section level1">
<h1>Input From Scratch</h1>
<p>In order to generate a <code>pagoo</code> object, the only mandatory data structure is a <code>data.frame</code> which has to contain 3 basic columns. Let’s illustrate it with an toy dataset.</p>
<p>First load <code>pagoo</code>:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true"></a><span class="kw">library</span>(pagoo) <span class="co"># Load package</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true"></a>tgz &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&#39;extdata&#39;</span>, <span class="st">&#39;toy_data.tar.gz&#39;</span>, <span class="dt">package =</span> <span class="st">&#39;pagoo&#39;</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true"></a><span class="kw">untar</span>(<span class="dt">tarfile =</span> tgz, <span class="dt">exdir =</span> <span class="kw">tempdir</span>()) <span class="co"># Decompress example dataset</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true"></a>files &lt;-<span class="st"> </span><span class="kw">list.files</span>(<span class="dt">path =</span> <span class="kw">tempdir</span>(), <span class="dt">full.names =</span> <span class="ot">TRUE</span>, <span class="dt">pattern =</span> <span class="st">&#39;tsv$|fasta$&#39;</span>) <span class="co"># List files</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true"></a>files</span></code></pre></div>
<pre><code>## [1] &quot;/tmp/Rtmpwl0FiU/case_clusters_meta.tsv&quot;
## [2] &quot;/tmp/Rtmpwl0FiU/case_df.tsv&quot;           
## [3] &quot;/tmp/Rtmpwl0FiU/case_orgs_meta.tsv&quot;    
## [4] &quot;/tmp/Rtmpwl0FiU/organismA.fasta&quot;       
## [5] &quot;/tmp/Rtmpwl0FiU/organismB.fasta&quot;       
## [6] &quot;/tmp/Rtmpwl0FiU/organismC.fasta&quot;       
## [7] &quot;/tmp/Rtmpwl0FiU/organismD.fasta&quot;       
## [8] &quot;/tmp/Rtmpwl0FiU/organismE.fasta&quot;</code></pre>
<p>The file we need now is <code>case_df.tsv</code>. Lets load it and see what’s its structure:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true"></a>data_file &lt;-<span class="st"> </span><span class="kw">grep</span>(<span class="st">&quot;case_df.tsv&quot;</span>, files, <span class="dt">value =</span> <span class="ot">TRUE</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true"></a>data &lt;-<span class="st"> </span><span class="kw">read.table</span>(data_file, <span class="dt">header =</span> <span class="ot">TRUE</span>, <span class="dt">sep =</span> <span class="st">&#39;</span><span class="ch">\t</span><span class="st">&#39;</span>, <span class="dt">quote =</span> <span class="st">&#39;&#39;</span>)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true"></a><span class="kw">head</span>(data)</span></code></pre></div>
<pre><code>##      gene       org cluster                             annot
## 1 gene081 organismA   OG001  Thioesterase superfamily protein
## 2 gene122 organismB   OG001          Thioesterase superfamily
## 3 gene299 organismC   OG001  Thioesterase superfamily protein
## 4 gene186 organismD   OG001  Thioesterase superfamily protein
## 5 gene076 organismE   OG001          Thioesterase superfamily
## 6 gene352 organismA   OG002 Inherit from proNOG: Thioesterase</code></pre>
<p>So it is a <code>data.frame</code> with 4 columns. The first one with the name of each gene, the second one with the organism to which each gene belongs, the third one with the cluster to which each gene was assigned in the pangenome reconstruction, and the last one with annotation metadata for each gene. Of the 4 columns, the former 3 are required, and <code>pagoo</code> will look for columns named “gene”, “org”, and “cluster”. More columns are optional, and you can add as many as you want (or none) to add metadata of each gene.</p>
<p>With only this data (even ignoring the fourth column, which is metadata), you can start working with pagoo:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true"></a>pg &lt;-<span class="st"> </span><span class="kw">pagoo</span>(<span class="dt">data =</span> data)</span></code></pre></div>
<p>but let’s continue adding more.</p>
<div id="including-organism-metadata" class="section level2">
<h2>Including organism metadata</h2>
<p>The next 2 <code>.tsv</code> files contains metadata for each cluster and for each organism, respectively, and are optional arguments.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true"></a><span class="co"># Organism metadata</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true"></a>orgs_file &lt;-<span class="st"> </span><span class="kw">grep</span>(<span class="st">&quot;case_orgs_meta.tsv&quot;</span>, files, <span class="dt">value =</span> <span class="ot">TRUE</span>)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true"></a>orgs_meta &lt;-<span class="st"> </span><span class="kw">read.table</span>(orgs_file, <span class="dt">header =</span> <span class="ot">TRUE</span>, <span class="dt">sep =</span> <span class="st">&#39;</span><span class="ch">\t</span><span class="st">&#39;</span>, <span class="dt">quote =</span> <span class="st">&#39;&#39;</span>)</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true"></a><span class="kw">head</span>(orgs_meta)</span></code></pre></div>
<pre><code>##         org sero  country
## 1 organismA    a Westeros
## 2 organismB    b Westeros
## 3 organismC    c Westeros
## 4 organismD    a    Essos
## 5 organismE    b    Essos</code></pre>
<p>On this <code>data.frame</code> we have a column named <code>org</code> which is mandatory in case you provide this argument. Other columns are metadata associated to each organism. Beware that organisms provided in this table (<code>orgs_meta$org</code>) must coincide with the names provided in the <code>data$org</code> field, in order to correctly map each variables.</p>
</div>
<div id="including-cluster-metadata" class="section level2">
<h2>Including cluster metadata</h2>
<p>Last file contains metadata associated with each cluster of orthologous:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true"></a><span class="co"># Cluster metadata</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true"></a>clust_file &lt;-<span class="st"> </span><span class="kw">grep</span>(<span class="st">&quot;case_clusters_meta.tsv&quot;</span>, files, <span class="dt">value =</span> <span class="ot">TRUE</span>)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true"></a>clust_meta &lt;-<span class="st"> </span><span class="kw">read.table</span>(clust_file, <span class="dt">header =</span> <span class="ot">TRUE</span>, <span class="dt">sep =</span> <span class="st">&#39;</span><span class="ch">\t</span><span class="st">&#39;</span>, <span class="dt">quote =</span> <span class="st">&#39;&#39;</span>)</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true"></a><span class="kw">head</span>(clust_meta)</span></code></pre></div>
<pre><code>##   cluster   kegg  cog
## 1   OG001   &lt;NA&gt;    S
## 2   OG002   &lt;NA&gt;    S
## 3   OG003   &lt;NA&gt; &lt;NA&gt;
## 4   OG004   &lt;NA&gt;    D
## 5   OG005 K01990    V
## 6   OG006   &lt;NA&gt;    V</code></pre>
<p>Again, the column <code>clust_meta$cluster</code> must contain the same elements as <code>data$cluster</code> column to be able to map one into the other.</p>
<p>With all this data the <code>pagoo</code> object will look much more complete. But you can still add sequence information to the pangenome, which makes it much more useful and interesting to work with.</p>
</div>
<div id="including-sequences" class="section level2">
<h2>Including sequences</h2>
<p>In this made up dataset we have 5 organisms, so if you decide to add sequences to the pangenome you must provide them for all 5 organisms. The type of data needed is a DNA multifasta file for each organism, in which each sequence is a gene whose name can be mapped to the <code>data$gene</code> column. You must first load the sequences into a <code>list</code>, and name each list element as the organism provided in <code>data$org</code> (as well as <code>org_meta$org</code>). The <code>list</code> would look something like:</p>
<ol style="list-style-type: decimal">
<li>organism1
<ul>
<li>gene1</li>
<li>gene2</li>
<li>…</li>
<li>geneN</li>
</ul></li>
<li>organism2
<ul>
<li>gene1</li>
<li>gene2</li>
<li>…</li>
<li>geneM</li>
</ul></li>
<li>organism3
<ul>
<li>gene1</li>
<li>gene2</li>
<li>…</li>
<li>geneP</li>
</ul></li>
<li>…</li>
</ol>
<p>In the case of the example we are working on:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true"></a><span class="co"># Sequences</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true"></a>fasta_files &lt;-<span class="st"> </span><span class="kw">grep</span>(<span class="st">&quot;[.]fasta&quot;</span>, files, <span class="dt">value =</span> <span class="ot">TRUE</span>) <span class="co"># List fasta files</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true"></a><span class="kw">names</span>(fasta_files) &lt;-<span class="st"> </span><span class="kw">sub</span>(<span class="st">&#39;[.]fasta&#39;</span>, <span class="st">&#39;&#39;</span>, <span class="kw">basename</span>(fasta_files)) <span class="co"># Name them </span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true"></a><span class="co"># Read fasta files with Biostrings:</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true"></a><span class="kw">library</span>(Biostrings)</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true"></a>sq &lt;-<span class="st"> </span><span class="kw">lapply</span>(fasta_files, readDNAStringSet)</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true"></a><span class="kw">class</span>(sq) <span class="co"># Is list?</span></span></code></pre></div>
<pre><code>## [1] &quot;list&quot;</code></pre>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true"></a><span class="kw">length</span>(sq) <span class="co"># One list element per organism</span></span></code></pre></div>
<pre><code>## [1] 5</code></pre>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true"></a><span class="kw">names</span>(sq) <span class="co"># Names are the same as in data$org</span></span></code></pre></div>
<pre><code>## [1] &quot;organismA&quot; &quot;organismB&quot; &quot;organismC&quot; &quot;organismD&quot; &quot;organismE&quot;</code></pre>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true"></a><span class="kw">class</span>(sq[[<span class="dv">1</span>]]) <span class="co"># Class of each element of the list</span></span></code></pre></div>
<pre><code>## [1] &quot;DNAStringSet&quot;
## attr(,&quot;package&quot;)
## [1] &quot;Biostrings&quot;</code></pre>
<p>And we have a <code>list</code> of <code>DNAStringSet</code> (Biostrings package). Now we can load a quite complete <code>pagoo</code> object (you could still add more metadata to genes, clusters, or organisms):</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true"></a>p &lt;-<span class="st"> </span><span class="kw">pagoo</span>(<span class="dt">data =</span> data, <span class="co"># Required data</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true"></a>           <span class="dt">org_meta =</span> orgs_meta, <span class="co"># Organism&#39;s metadata</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true"></a>           <span class="dt">cluster_meta =</span> clust_meta, <span class="co"># Cluster&#39;s metadata</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true"></a>           <span class="dt">sequences =</span> sq) <span class="co"># Sequences</span></span></code></pre></div>
</div>
</div>
<div id="input-from-pangenome-reconstruction-software" class="section level1">
<h1>Input From Pangenome Reconstruction Software</h1>
<p>All the above stuff with preparing data and loading classes seems difficult and time-consuming, but in real life working datasets this will be rarely needed. We are explaining it here to provide full details about how the software works, but this package also provides functions to automatically read-in output files from pangenome reconstruction software into <code>pagoo</code>, avoiding any formatting or manipulation of data.</p>
<p>Currently <code>pagoo</code> supports input from <a href="https://sanger-pathogens.github.io/Roary/">roary</a> (Page et al., 2015), which has been the standard and most cited software for pangenome reconstruction, and <a href="https://github.com/gtonkinhill/panaroo">panaroo</a> (Tonkin-Hill et al., 2020). It is worth noticing that as roary become the most used software in this field, other tools as <a href="https://github.com/zheminzhou/PEPPAN">PEPPAN</a>, <a href="https://github.com/SionBayliss/PIRATE">PRIATE</a> and panaroo include scripts to convert their output to roary’s format. To work with <code>roary</code>’s output, please refer to <code>?roary_2_pagoo</code> documentation. Although panaroo also includes a tool for this, we provide a function to load their native output format, see <code>?panaroo_2_pagoo</code>. For both functions you will only need the <code>.gff</code> files used as input for roary of panaroo, and the <code>gene_presence_absence.csv</code> file.</p>
<p>Also, we have created our own pangenome reconstruction software called <a href="https://github.com/iferres/pewit">pewit</a> (Ferrés et al., still unpublished), which automatically generates a <code>pagoo</code>-like object to perform downstream analyses. This object contain all the methods and fields <code>pagoo</code> provides, plus a set of methods and fields exclusive to this software.</p>
<p>Other good pangenome reconstruction software already exists like <a href="https://pangenome.org/">PanX</a> (Ding et al., 2018), <a href="https://github.com/cran/micropan">micropan</a> (Snipen &amp; Liland, 2015), <a href="https://github.com/eead-csic-compbio/get_homologues">GET_HOMOLOGUES</a> (Contreras-Moreira &amp; Vinuesa, 2013), among others. We plan to provide support to some of them in the future.</p>
</div>
<div id="adding-metadata-after-object-creation" class="section level1">
<h1>Adding Metadata After Object Creation</h1>
<p>After object creation, you may want to add new metadata given new information or as result of posterior analyses. <code>pagoo</code> objects include a function to add columns of metadata to each gene, each cluster, or each organism. To illustrate it, we will add a new column to the <code>$organisms</code> field named <code>host</code> to add made up information about the host where each genome was isolated from.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true"></a>host_df &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">org =</span> p<span class="op">$</span>organisms<span class="op">$</span>org, <span class="dt">host =</span> <span class="kw">c</span>(<span class="st">&quot;Cow&quot;</span>, <span class="st">&quot;Dog&quot;</span>, <span class="st">&quot;Cat&quot;</span>, <span class="st">&quot;Cow&quot;</span>, <span class="st">&quot;Sheep&quot;</span>))</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true"></a>p<span class="op">$</span><span class="kw">add_metadata</span>(<span class="dt">map =</span> <span class="st">&quot;org&quot;</span>, host_df)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true"></a>p<span class="op">$</span>organisms</span></code></pre></div>
<pre><code>## DataFrame with 5 rows and 4 columns
##         org        sero     country        host
##    &lt;factor&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt;
## 1 organismA           a    Westeros         Cow
## 2 organismB           b    Westeros         Dog
## 3 organismC           c    Westeros         Cat
## 4 organismD           a       Essos         Cow
## 5 organismE           b       Essos       Sheep</code></pre>
<p>In order to allow <code>pagoo</code> correct data mapping, the values in the first column of the metadata table should be available at <code>p$organisms$org</code>, and its column header must also be named <code>org</code>.</p>
<p>As said, you can add <code>gene</code> or <code>cluster</code> metadata following the same idea.</p>
</div>
<div id="saving-and-loading-pagoo-objects" class="section level1">
<h1>Saving and Loading <code>pagoo</code> objects</h1>
<p>Once loaded, it has two methods for saving and reloading to a new R session. The first one is by saving them as flat (text) files:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true"></a>tmp &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="kw">tempdir</span>(), <span class="st">&quot;pangenome&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;/&quot;</span>)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true"></a>p<span class="op">$</span><span class="kw">write_pangenome</span>(<span class="dt">dir =</span> tmp)</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true"></a><span class="kw">list.files</span>(tmp, <span class="dt">full.names =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<pre><code>## [1] &quot;/tmp/Rtmpwl0FiU/pangenome/clusters.tsv&quot; 
## [2] &quot;/tmp/Rtmpwl0FiU/pangenome/data.tsv&quot;     
## [3] &quot;/tmp/Rtmpwl0FiU/pangenome/organisms.tsv&quot;</code></pre>
<p>This creates a directory with 3 text files. The advantage of this approach is that you can analyze it outside R, the disadvantage is that from a reproducibility point of view reading text could be less stable since class or number precision can be lost, and also you can’t save the state of the object in any given time. Only available organisms/genes/clusters are saved, and if you reload the class using the tsv files, any previously dropped organisms/gene/cluster won’t be available any more. (For information about dropping/recovering organisms, see “Subset” tutorial).</p>
<p>If you want to save the object and continue working with it in other R session, we recommend to save them as R objects with the RDS methods provided:</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true"></a>rds &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="kw">tempdir</span>(), <span class="st">&quot;pangenome.RDS&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;/&quot;</span>)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true"></a>p<span class="op">$</span><span class="kw">save_pangenomeRDS</span>(<span class="dt">file =</span> rds)</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true"></a>p2 &lt;-<span class="st"> </span><span class="kw">load_pangenomeRDS</span>(rds)</span></code></pre></div>
<p>This method is more stable (compatible between <code>pagoo</code> versions), secure (uses the same metadata classes, and precision isn’t lost), and convenient (the exact state of the object saved is restored, keeping dropped organisms/genes/clusters hidden, available to be recovered, and other object state configuration, e.g. <code>core_level</code>, is also saved).</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
