<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Subseting</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>
<script>// Hide empty <a> tag within highlighted CodeBlock for screen reader accessibility (see https://github.com/jgm/pandoc/issues/6352#issuecomment-626106786) -->
// v0.0.1
// Written by JooYoung Seo (jooyoung@psu.edu) and Atsushi Yasumoto on June 1st, 2020.

document.addEventListener('DOMContentLoaded', function() {
  const codeList = document.getElementsByClassName("sourceCode");
  for (var i = 0; i < codeList.length; i++) {
    var linkList = codeList[i].getElementsByTagName('a');
    for (var j = 0; j < linkList.length; j++) {
      if (linkList[j].innerHTML === "") {
        linkList[j].setAttribute('aria-hidden', 'true');
      }
    }
  }
});
</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Subseting</h1>



<p>One fundamental operation when working with pangenomic data is to subset it in order to better analyze it. <code>pagoo</code> provides three convenient ways that can help in different scenarios. First we will see predefined subsets, then we will understand how to use the classic R’s <code>[</code> operator with <code>pagoo</code> objects, and finally we will see how to temporarily remove one or more organisms from the dataset, and then to recover them.</p>
<p>Start the tutorial by loading the included <em>Campylobacter</em> pangenome.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true"></a><span class="kw">library</span>(pagoo) <span class="co"># Load package</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true"></a>toy_rds &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&#39;extdata&#39;</span>, <span class="st">&#39;campylobacter.RDS&#39;</span>, <span class="dt">package =</span> <span class="st">&#39;pagoo&#39;</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true"></a>p &lt;-<span class="st"> </span><span class="kw">load_pangenomeRDS</span>(toy_rds)</span></code></pre></div>
<div id="predefined-subsets" class="section level1">
<h1>Predefined Subsets</h1>
<p>Predefined subsets were already superficially seen in <strong>3 - Querying Data</strong> tutorial. Here we will review them and understand its notation.</p>
<table>
<thead>
<tr class="header">
<th>-</th>
<th><code>$core_*</code></th>
<th><code>$shell_*</code></th>
<th><code>$cloud_*</code></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong><code>$*_genes</code></strong></td>
<td><code>$core_genes</code></td>
<td><code>$shell_genes</code></td>
<td><code>$cloud_genes</code></td>
</tr>
<tr class="even">
<td><strong><code>$*_clusters</code></strong></td>
<td><code>$core_clusters</code></td>
<td><code>$shell_clusters</code></td>
<td><code>$cloud_clusters</code></td>
</tr>
<tr class="odd">
<td><strong><code>$*_sequences</code></strong></td>
<td><code>$core_sequences</code></td>
<td><code>$shell_sequences</code></td>
<td><code>$cloud_sequences</code></td>
</tr>
</tbody>
</table>
<p>As seen in the above table, the notation is quite straightforward. Lets take <code>clusters</code> for instance to illustrate it better.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true"></a>p<span class="op">$</span>clusters</span></code></pre></div>
<pre><code>## DataFrame with 2588 rows and 2 columns
##        cluster              Pfam_Arch
##       &lt;factor&gt;            &lt;character&gt;
## 1    group0001           2-Hacid_dh_C
## 2    group0002 2-Hacid_dh_C;2-Hacid..
## 3    group0003 2-Hacid_dh_C;ACT;2-H..
## 4    group0004        2Fe-2S_thioredx
## 5    group0005    4HB_MCP_1;MCPsignal
## ...        ...                    ...
## 2584 group2584              zf-RING_7
## 2585 group2585               zf-TFIIB
## 2586 group2586                   ZinT
## 2587 group2587                   ZnuA
## 2588 group2588               ZT_dimer</code></pre>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true"></a>p<span class="op">$</span>core_clusters</span></code></pre></div>
<pre><code>## DataFrame with 1627 rows and 2 columns
##        cluster              Pfam_Arch
##       &lt;factor&gt;            &lt;character&gt;
## 1    group0001           2-Hacid_dh_C
## 2    group0002 2-Hacid_dh_C;2-Hacid..
## 3    group0003 2-Hacid_dh_C;ACT;2-H..
## 4    group0004        2Fe-2S_thioredx
## 5    group0006    4HB_MCP_1;MCPsignal
## ...        ...                    ...
## 1623 group2584              zf-RING_7
## 1624 group2585               zf-TFIIB
## 1625 group2586                   ZinT
## 1626 group2587                   ZnuA
## 1627 group2588               ZT_dimer</code></pre>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true"></a>p<span class="op">$</span>shell_clusters</span></code></pre></div>
<pre><code>## DataFrame with 413 rows and 2 columns
##       cluster            Pfam_Arch
##      &lt;factor&gt;          &lt;character&gt;
## 1   group0005  4HB_MCP_1;MCPsignal
## 2   group0009        5_nucleotid_C
## 3   group0015                  AAA
## 4   group0019               AAA_22
## 5   group0021               AAA_22
## ...       ...                  ...
## 409 group2537 VirB3;CagE_TrbE_VirB
## 410 group2541                VirB8
## 411 group2543                VirB8
## 412 group2554               Y1_Tnp
## 413 group2557           YafQ_toxin</code></pre>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true"></a>p<span class="op">$</span>cloud_clusters</span></code></pre></div>
<pre><code>## DataFrame with 548 rows and 2 columns
##       cluster   Pfam_Arch
##      &lt;factor&gt; &lt;character&gt;
## 1   group0018      AAA_22
## 2   group0020      AAA_22
## 3   group0024      AAA_25
## 4   group0025      AAA_25
## 5   group0028      AAA_28
## ...       ...         ...
## 544 group2555  YafQ_toxin
## 545 group2566      YhcG_C
## 546 group2573        YopX
## 547 group2578  Zeta_toxin
## 548 group2579  Zeta_toxin</code></pre>
</div>
<div id="the-operator" class="section level1">
<h1>The <code>[</code> operator</h1>
<p>Using <code>[</code> for subset vectors, lists, or matrices in R is one of the most common operations we, as R users, do every day. Methods using <code>[</code> for subsetting can be divided into 2 types:</p>
<ol style="list-style-type: decimal">
<li>Subsetting vector or list like objects.</li>
<li>Subsetting matrix-like objects.</li>
</ol>
<p>In the following subsections we will see how to use them to subset <code>$pan_matrix</code>, <code>$genes</code>, <code>$clusters</code>, <code>$sequences</code>, or<code>$organisms</code> fields.</p>
<div id="vectorlist-notation" class="section level2">
<h2>Vector/List notation</h2>
<p>When we subset a vector or list in R we use a single number or a vector of numbers, lets say from <code>i</code> to <code>j</code>, to pick those elements by indexes: <code>x[i:j]</code>. In <code>pagoo</code> we implement a method for this generic function to subset data fields, where indexes represent <strong>clusters</strong>. So instead of subsetting directly the final object (which you can of course) you apply this method directly to the object, and then select any of the data fields.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true"></a><span class="co"># From clusters 1 to 3</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true"></a>p[<span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">5</span>, <span class="dv">18</span>)]<span class="op">$</span>pan_matrix</span></code></pre></div>
<pre><code>##            group0001 group0005 group0018
## 16244_6_6          1         1         0
## 16244_6_18         1         1         0
## 17059_2_16         1         1         0
## 17059_2_23         1         0         0
## 17059_2_27         1         1         0
## 17150_1_73         1         1         0
## 17059_2_42         1         0         1</code></pre>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true"></a>p[<span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">5</span>, <span class="dv">18</span>)]<span class="op">$</span>genes</span></code></pre></div>
<pre><code>## SplitDataFrameList of length 3
## $group0001
## DataFrame with 7 rows and 10 columns
##     cluster        org             gene                    gid    geneName
##    &lt;factor&gt;   &lt;factor&gt;         &lt;factor&gt;            &lt;character&gt; &lt;character&gt;
## 1 group0001 16244_6_6  16244_6_6_00150  16244_6_6__16244_6_6..        ilvC
## 2 group0001 16244_6_18 16244_6_18_00172 16244_6_18__16244_6_..        ilvC
## 3 group0001 17059_2_16 17059_2_16_01012 17059_2_16__17059_2_..        ilvC
## 4 group0001 17059_2_23 17059_2_23_01040 17059_2_23__17059_2_..        ilvC
## 5 group0001 17059_2_27 17059_2_27_00492 17059_2_27__17059_2_..        ilvC
## 6 group0001 17150_1_73 17150_1_73_00221 17150_1_73__17150_1_..        ilvC
## 7 group0001 17059_2_42 17059_2_42_00176 17059_2_42__17059_2_..        ilvC
##                  product                 contig      from        to      strand
##              &lt;character&gt;            &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;character&gt;
## 1 ketol-acid reductois.. ERS672247|SC|contig0..    137333    138355           +
## 2 ketol-acid reductois.. ERS672259|SC|contig0..    173541    174563           +
## 3 ketol-acid reductois.. ERS739235|SC|contig0..     43645     44667           -
## 4 ketol-acid reductois.. ERS739242|SC|contig0..    173578    174600           +
## 5 ketol-acid reductois.. ERS739246|SC|contig0..    184035    185057           -
## 6 ketol-acid reductois.. ERS686652|SC|contig0..    207206    208228           +
## 7 ketol-acid reductois.. ERS739261|SC|contig0..    173569    174591           +
## 
## $group0005
## DataFrame with 5 rows and 10 columns
##     cluster        org             gene                    gid    geneName
##    &lt;factor&gt;   &lt;factor&gt;         &lt;factor&gt;            &lt;character&gt; &lt;character&gt;
## 1 group0005 16244_6_6  16244_6_6_00049  16244_6_6__16244_6_6..      mcpB_1
## 2 group0005 16244_6_18 16244_6_18_00072 16244_6_18__16244_6_..      mcpB_1
## 3 group0005 17059_2_16 17059_2_16_01270 17059_2_16__17059_2_..      mcpB_3
## 4 group0005 17059_2_27 17059_2_27_00592 17059_2_27__17059_2_..      mcpB_1
## 5 group0005 17150_1_73 17150_1_73_00073 17150_1_73__17150_1_..      mcpB_1
##                  product                 contig      from        to      strand
##              &lt;character&gt;            &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;character&gt;
## 1 methyl-accepting che.. ERS672247|SC|contig0..     46407     48068           +
## 2 methyl-accepting che.. ERS672259|SC|contig0..     82617     84278           +
## 3 methyl-accepting che.. ERS739235|SC|contig0..     82775     84436           +
## 4 methyl-accepting che.. ERS739246|SC|contig0..    274321    275982           -
## 5 methyl-accepting che.. ERS686652|SC|contig0..     82648     84309           +
## 
## $group0018
## DataFrame with 1 row and 10 columns
##     cluster        org             gene                    gid    geneName
##    &lt;factor&gt;   &lt;factor&gt;         &lt;factor&gt;            &lt;character&gt; &lt;character&gt;
## 1 group0018 17059_2_42 17059_2_42_00333 17059_2_42__17059_2_..            
##                  product                 contig      from        to      strand
##              &lt;character&gt;            &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;character&gt;
## 1 bacteriocin,putative.. ERS739261|SC|contig0..    303996    304949           -</code></pre>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true"></a>p[<span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">5</span>, <span class="dv">18</span>)]<span class="op">$</span>clusters</span></code></pre></div>
<pre><code>## DataFrame with 3 rows and 2 columns
##     cluster           Pfam_Arch
##    &lt;factor&gt;         &lt;character&gt;
## 1 group0001        2-Hacid_dh_C
## 2 group0005 4HB_MCP_1;MCPsignal
## 3 group0018              AAA_22</code></pre>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true"></a>p[<span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">5</span>, <span class="dv">18</span>)]<span class="op">$</span>sequences</span></code></pre></div>
<pre><code>## DNAStringSetList of length 3
## [[&quot;group0001&quot;]] 16244_6_18__16244_6_18_00172=ATGGCGATAACAGTTTATTACGACAAAGATTG...
## [[&quot;group0005&quot;]] 16244_6_18__16244_6_18_00072=ATGTCAAATTTAACTACTAATTTAACTACCAA...
## [[&quot;group0018&quot;]] 17059_2_42__17059_2_42_00333=ATGAGTAAATTTGAAGATATTAGGAATGAGCT...</code></pre>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true"></a>p[<span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">5</span>, <span class="dv">18</span>)]<span class="op">$</span>organisms</span></code></pre></div>
<pre><code>## DataFrame with 7 rows and 8 columns
##          org          id      strain      year     country        host
##     &lt;factor&gt; &lt;character&gt; &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;character&gt;
## 1 16244_6_6         FR15   2008/170h      2008      France       Human
## 2 16244_6_18        FR27   2012/185h      2012      France       Human
## 3 17059_2_16         AR1      99/801      1999   Argentina      Bovine
## 4 17059_2_23         AR8      04/875      2004   Argentina      Bovine
## 5 17059_2_27        AR12      06/195      2006   Argentina      Bovine
## 6 17150_1_73         CA1   001A-0374      2005      Canada       Human
## 7 17059_2_42         TW6        1830      2008      Taiwan       Human
##        source   accession
##   &lt;character&gt; &lt;character&gt;
## 1       Feces   ERS672247
## 2       Blood   ERS672259
## 3     Prepuce   ERS739235
## 4       Fetus   ERS739242
## 5          VM   ERS739246
## 6       Blood   ERS686652
## 7       Blood   ERS739261</code></pre>
<p>Note that in each case, indexes can be interpreted as clusters. In <code>pan_matrix</code> they refer to columns, in <code>genes</code> and <code>sequences</code> to elements in a list, in <code>clusters</code> rows in the dataframe, and in the case of <code>$organisms</code> it returns the organisms (rows) where those clusters are present. See what happen for this last case when we query for organisms present for a shell cluster:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true"></a>shell_clust &lt;-<span class="st"> </span>p<span class="op">$</span>shell_clusters<span class="op">$</span>cluster[<span class="dv">1</span>] <span class="co"># [1] &quot;OG0005&quot;</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true"></a>p[shell_clust]<span class="op">$</span>organisms</span></code></pre></div>
<pre><code>## DataFrame with 5 rows and 8 columns
##          org          id      strain      year     country        host
##     &lt;factor&gt; &lt;character&gt; &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;character&gt;
## 1 16244_6_6         FR15   2008/170h      2008      France       Human
## 2 16244_6_18        FR27   2012/185h      2012      France       Human
## 3 17059_2_16         AR1      99/801      1999   Argentina      Bovine
## 5 17059_2_27        AR12      06/195      2006   Argentina      Bovine
## 6 17150_1_73         CA1   001A-0374      2005      Canada       Human
##        source   accession
##   &lt;character&gt; &lt;character&gt;
## 1       Feces   ERS672247
## 2       Blood   ERS672259
## 3     Prepuce   ERS739235
## 5          VM   ERS739246
## 6       Blood   ERS686652</code></pre>
<p>Only (6) organisms which contain that shell cluster will be listed.</p>
</div>
<div id="matrix-notation" class="section level2">
<h2>Matrix notation</h2>
<p>The other use of <code>[</code> notation is when we subset a matrix-like object. In this case we provide 2 sets of indexes: the first for rows, and the second for columns, separated by a coma. <code>pagoo</code> interprets these indexes as organisms (rows) and clusters (columns), so you are referencing a set of genes (cell value). Basically you are referencing cells in the pan_matrix.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true"></a>p<span class="op">$</span>pan_matrix[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>, <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">5</span>, <span class="dv">18</span>)]</span></code></pre></div>
<pre><code>##            group0001 group0005 group0018
## 16244_6_6          1         1         0
## 16244_6_18         1         1         0
## 17059_2_16         1         1         0</code></pre>
<p>Here we are selecting the first 3 organisms, and the clusters indexed as 1, 5, and 18. Note that we see (sum columns) 3 genes in the first cluster, 3 in the second, and 0 in the third. Now we flip the notation and subset directly the <code>pagoo</code> object, and then ask for a data field:</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true"></a>p[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>, <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">5</span>, <span class="dv">18</span>)]<span class="op">$</span>pan_matrix <span class="co"># The same as above</span></span></code></pre></div>
<pre><code>##            group0001 group0005 group0018
## 16244_6_6          1         1         0
## 16244_6_18         1         1         0
## 17059_2_16         1         1         0</code></pre>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true"></a>p[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>, <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">5</span>, <span class="dv">18</span>)]<span class="op">$</span>genes</span></code></pre></div>
<pre><code>## SplitDataFrameList of length 2
## $group0001
## DataFrame with 3 rows and 10 columns
##     cluster        org             gene                    gid    geneName
##    &lt;factor&gt;   &lt;factor&gt;         &lt;factor&gt;            &lt;character&gt; &lt;character&gt;
## 1 group0001 16244_6_6  16244_6_6_00150  16244_6_6__16244_6_6..        ilvC
## 2 group0001 16244_6_18 16244_6_18_00172 16244_6_18__16244_6_..        ilvC
## 3 group0001 17059_2_16 17059_2_16_01012 17059_2_16__17059_2_..        ilvC
##                  product                 contig      from        to      strand
##              &lt;character&gt;            &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;character&gt;
## 1 ketol-acid reductois.. ERS672247|SC|contig0..    137333    138355           +
## 2 ketol-acid reductois.. ERS672259|SC|contig0..    173541    174563           +
## 3 ketol-acid reductois.. ERS739235|SC|contig0..     43645     44667           -
## 
## $group0005
## DataFrame with 3 rows and 10 columns
##     cluster        org             gene                    gid    geneName
##    &lt;factor&gt;   &lt;factor&gt;         &lt;factor&gt;            &lt;character&gt; &lt;character&gt;
## 1 group0005 16244_6_6  16244_6_6_00049  16244_6_6__16244_6_6..      mcpB_1
## 2 group0005 16244_6_18 16244_6_18_00072 16244_6_18__16244_6_..      mcpB_1
## 3 group0005 17059_2_16 17059_2_16_01270 17059_2_16__17059_2_..      mcpB_3
##                  product                 contig      from        to      strand
##              &lt;character&gt;            &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;character&gt;
## 1 methyl-accepting che.. ERS672247|SC|contig0..     46407     48068           +
## 2 methyl-accepting che.. ERS672259|SC|contig0..     82617     84278           +
## 3 methyl-accepting che.. ERS739235|SC|contig0..     82775     84436           +</code></pre>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true"></a>p[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>, <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">5</span>, <span class="dv">18</span>)]<span class="op">$</span>clusters</span></code></pre></div>
<pre><code>## DataFrame with 2 rows and 2 columns
##     cluster           Pfam_Arch
##    &lt;factor&gt;         &lt;character&gt;
## 1 group0001        2-Hacid_dh_C
## 2 group0005 4HB_MCP_1;MCPsignal</code></pre>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true"></a>p[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>, <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">5</span>, <span class="dv">18</span>)]<span class="op">$</span>sequences</span></code></pre></div>
<pre><code>## DNAStringSetList of length 2
## [[&quot;group0001&quot;]] 16244_6_18__16244_6_18_00172=ATGGCGATAACAGTTTATTACGACAAAGATTG...
## [[&quot;group0005&quot;]] 16244_6_18__16244_6_18_00072=ATGTCAAATTTAACTACTAATTTAACTACCAA...</code></pre>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true"></a>p[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>, <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">5</span>, <span class="dv">18</span>)]<span class="op">$</span>organisms</span></code></pre></div>
<pre><code>## DataFrame with 3 rows and 8 columns
##          org          id      strain      year     country        host
##     &lt;factor&gt; &lt;character&gt; &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;character&gt;
## 1 16244_6_6         FR15   2008/170h      2008      France       Human
## 2 16244_6_18        FR27   2012/185h      2012      France       Human
## 3 17059_2_16         AR1      99/801      1999   Argentina      Bovine
##        source   accession
##   &lt;character&gt; &lt;character&gt;
## 1       Feces   ERS672247
## 2       Blood   ERS672259
## 3     Prepuce   ERS739235</code></pre>
<p>So this selection returns a list of length 2, with 3 elements each for <code>$genes</code> and <code>$sequences</code>, a dataframe showing only the selected clusters in the <code>$clusters</code> field, and a dataframe showing only the selected organisms in the <code>$organisms</code> field.</p>
<p>We found this implementation quite useful for both data exploration and fine grained analysis.</p>
</div>
</div>
<div id="dropping-and-recovering-organisms" class="section level1">
<h1>Dropping and Recovering Organisms</h1>
<p>One useful feature implemented in <code>pagoo</code> is the possibility of easily removing (hiding) organisms from the dataset. This is useful if during the analysis we identify some genome with weird characteristics (i.e. potentially contaminated), or if we want to focus just in a subset of genomes of interest given any metadata value, or if we included an outgroup for phylogenetic purposes but we want to remove it from downstream analyses. Let’s see how it works..</p>
<div id="dropping-organisms" class="section level2">
<h2>Dropping organisms</h2>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true"></a>p<span class="op">$</span>organisms</span></code></pre></div>
<pre><code>## DataFrame with 7 rows and 8 columns
##          org          id      strain      year     country        host
##     &lt;factor&gt; &lt;character&gt; &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;character&gt;
## 1 16244_6_6         FR15   2008/170h      2008      France       Human
## 2 16244_6_18        FR27   2012/185h      2012      France       Human
## 3 17059_2_16         AR1      99/801      1999   Argentina      Bovine
## 4 17059_2_23         AR8      04/875      2004   Argentina      Bovine
## 5 17059_2_27        AR12      06/195      2006   Argentina      Bovine
## 6 17150_1_73         CA1   001A-0374      2005      Canada       Human
## 7 17059_2_42         TW6        1830      2008      Taiwan       Human
##        source   accession
##   &lt;character&gt; &lt;character&gt;
## 1       Feces   ERS672247
## 2       Blood   ERS672259
## 3     Prepuce   ERS739235
## 4       Fetus   ERS739242
## 5          VM   ERS739246
## 6       Blood   ERS686652
## 7       Blood   ERS739261</code></pre>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true"></a>p<span class="op">$</span>summary_stats</span></code></pre></div>
<pre><code>## DataFrame with 4 rows and 2 columns
##      Category    Number
##   &lt;character&gt; &lt;integer&gt;
## 1       Total      2588
## 2        Core      1627
## 3       Shell       413
## 4       Cloud       548</code></pre>
<p>We can see that we have 7 organisms, and we have the pangenome summary statistics. Let’s say, for instance, that we want to exclude french isolates (because any reason) from the dataset, which are indexed 1 and 2 (rows) in the above dataframe.</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true"></a>p<span class="op">$</span><span class="kw">drop</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>) </span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true"></a>p<span class="op">$</span>organisms <span class="co"># Updated organisms !!</span></span></code></pre></div>
<pre><code>## DataFrame with 5 rows and 8 columns
##          org          id      strain      year     country        host
##     &lt;factor&gt; &lt;character&gt; &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;character&gt;
## 3 17059_2_16         AR1      99/801      1999   Argentina      Bovine
## 4 17059_2_23         AR8      04/875      2004   Argentina      Bovine
## 5 17059_2_27        AR12      06/195      2006   Argentina      Bovine
## 6 17150_1_73         CA1   001A-0374      2005      Canada       Human
## 7 17059_2_42         TW6        1830      2008      Taiwan       Human
##        source   accession
##   &lt;character&gt; &lt;character&gt;
## 3     Prepuce   ERS739235
## 4       Fetus   ERS739242
## 5          VM   ERS739246
## 6       Blood   ERS686652
## 7       Blood   ERS739261</code></pre>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true"></a>p<span class="op">$</span>summary_stats <span class="co"># Updated stats !!</span></span></code></pre></div>
<pre><code>## DataFrame with 4 rows and 2 columns
##      Category    Number
##   &lt;character&gt; &lt;integer&gt;
## 1       Total      2504
## 2        Core      1648
## 3       Shell       309
## 4       Cloud       547</code></pre>
<p>As you see, not only the french organisms have been removed from the dataset, but also summary statistics have been updated! This is also true for any other data field: <code>$pan_matrix</code>, <code>$genes</code>, <code>$clusters</code>, <code>$sequences</code>, and<code>$organisms</code> fields are automatically updated. Also any embedded statistical or visualization methods will now only consider available organisms/clusters/genes. When we drop an organisms, we are hiding all features associated to that genomes including genes, sequences, gene clusters and metadata.</p>
<p><strong>It’s important to note that you don’t have to reassign the object to a new one, it is self modified (in place modification). That’s R6 reference semantics (see R6 documentation for details), use with caution.</strong></p>
</div>
<div id="recovering-dropped-organisms" class="section level2">
<h2>Recovering dropped organisms</h2>
<p>You can recover dropped organisms. To see if we have any hidden organism, we use the <code>$dropped</code> field, and then recover it using its index or its name.</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true"></a>p<span class="op">$</span>dropped</span></code></pre></div>
<pre><code>##            1            2 
##  &quot;16244_6_6&quot; &quot;16244_6_18&quot;</code></pre>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true"></a>p<span class="op">$</span><span class="kw">recover</span>( p<span class="op">$</span>dropped )</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true"></a>p<span class="op">$</span>organisms</span></code></pre></div>
<pre><code>## DataFrame with 7 rows and 8 columns
##          org          id      strain      year     country        host
##     &lt;factor&gt; &lt;character&gt; &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;character&gt;
## 1 16244_6_6         FR15   2008/170h      2008      France       Human
## 2 16244_6_18        FR27   2012/185h      2012      France       Human
## 3 17059_2_16         AR1      99/801      1999   Argentina      Bovine
## 4 17059_2_23         AR8      04/875      2004   Argentina      Bovine
## 5 17059_2_27        AR12      06/195      2006   Argentina      Bovine
## 6 17150_1_73         CA1   001A-0374      2005      Canada       Human
## 7 17059_2_42         TW6        1830      2008      Taiwan       Human
##        source   accession
##   &lt;character&gt; &lt;character&gt;
## 1       Feces   ERS672247
## 2       Blood   ERS672259
## 3     Prepuce   ERS739235
## 4       Fetus   ERS739242
## 5          VM   ERS739246
## 6       Blood   ERS686652
## 7       Blood   ERS739261</code></pre>
<p>Now french isolates and all its features are again available.</p>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
